js/base-mod.js

```diff
@@ -4,12 +4,34 @@
 function d20plusMod () {
 	d20plus.mod = {};
 
-	/* eslint-disable */
+	d20plus.mod.setMode = function (t) {
+		d20plus.ut.log(`Setting mode ${t}`);
+		const preserveDrawingColor = (stash) => {
+			const drawingTools = ["rect", "ellipse", "text", "path", "polygon"];
+			const drawingProps = [{nm: "fill", el: "fillcolor"}, {nm: "color", el: "strokecolor"}];
+			if (!drawingTools.includes(t)) return;
+			drawingProps.forEach(prop => {
+				if (stash) d20plus.mod[`drawing${prop.nm}`] = d20.engine.canvas.freeDrawingBrush[prop.nm];
+				else {
+					if (d20plus.mod[`drawingcolor`] === "rgb(0, 0, 0)" || !d20plus.mod[`drawingcolor`]) return;
+					$(`#path_${prop.el}`).val(d20plus.mod[`drawing${prop.nm}`]).trigger("change");
+				}
+			});
+		}
+		try {
+			preserveDrawingColor(true);
+			d20.Campaign.activePage().setModeRef(t);
+			preserveDrawingColor();
+		} catch (e) {
+			d20plus.ut.log(`Switching using legacy because ${e.message}`);
+			d20plus.mod.setModeLegacy(t);
```

Comment:
q without reading the code: do we need to call `preserveDrawingColor();` after this?

Reply:
`preserveDrawingColor();` contains all logic related to colors. On first call it stashes the colors, on second call it applies them after the tool has been selected. `setModeLegacy()` has its own drawing color preservation.